### PR TITLE
Implemented the GPU option in a full virtual machine solution.

### PR DESCRIPTION
**Pull request by Mahmoud-Emad**
### Description

Added two select areas to the node GPU cards

### Changes

List of changes this PR includes

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/638

### Screenshots/Videos
- [Screencast from 22-06-23 19:11:45.webm](https://github.com/threefoldtech/tfgrid-sdk-ts/assets/57001890/3ce98663-af55-4fcd-9fe5-eacf4d947d27)

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

File: packages/playground/src/components/select_gpu_node.vue [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+              envs: [{ key: "SSH_KEY", value: profileManager.profile!.ssh }],
+              rootFilesystemSize: 2,
+              hasGPU: filters.hasGPU,
+              rentedBy: filters.hasGPU ? profileManager.profile!.twinId : undefined,
+            },
+          ],
+          network: { addAccess: filters.wireguard },
+        })
+        .then(res =>
+          res.forEach(nodes => {
```
Comment (MohamedElmdary) [inline packages/playground/src/components/select_gpu_node.vue:166]: Please catch the possible errors and handle them by showing them to the user under the vuetify input

Comment (Mahmoud-Emad) [reply packages/playground/src/components/select_gpu_node.vue]: Done updated

File: packages/playground/src/weblets/full_vm.vue [hunk 2]
```diff
@@ -118,10 +118,21 @@
           <v-switch color="primary" inset label="Add Wireguard Access" v-model="wireguard" />
         </input-tooltip>
 
+        <input-tooltip
+          inline
+          tooltip="
+          Selecting a Node with GPU: Important Note
+          When selecting a node with GPU resources, please ensure that you have rented the node. To rent a node and gain access to GPU capabilities, you can utilize our dashboard.
```
Comment (AhmedHanafy725) [inline packages/playground/src/weblets/full_vm.vue:125]:
```suggestion
          Selecting a Node with GPU.
          When selecting a node with GPU resources, please make sure that you have a rented node. To rent a node and gain access to GPU capabilities, you can use our dashboard.
```

File: packages/stats/src/components/CardList.svelte [hunk 3]
```diff
@@ -64,9 +64,9 @@
       },
       {
         id: 7,
-        data: data.accessNodes,
-        title: "Access Nodes",
-        icon: "/assets/access.svg",
+        data: data.gpus,
+        title: "GPUs",
+        icon: "/assets/gpu.svg",
```
Comment (AhmedHanafy725) [inline packages/stats/src/components/CardList.svelte:69]: why these changes are here?

File: packages/stats/src/utils/fetchData.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@ import axios from "axios";
 import { writable } from "svelte/store";
 
 export interface IStatsRes {
+  gpus: number;
```
Comment (AhmedHanafy725) [inline packages/stats/src/utils/fetchData.ts:5]: this one too

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

File: packages/playground/src/components/select_gpu_node.vue [hunk 5]
```diff
@@ -0,0 +1,188 @@
+<template>
+  <section>
+    <h6 class="text-h5 mb-4">Select a node</h6>
+    <v-alert class="mb-2" type="warning" variant="tonal" v-if="!loadingNodes && selectedNode === undefined">
+      Seems to be there are no nodes rented by you that match your selected resources, try to reduce/change you'r
+      resources and try again.
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/components/select_gpu_node.vue:6]: updated

File: packages/playground/src/components/select_gpu_node.vue [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+      v-if="selectedNode"
+      :rules="[
+        validators.required('Please select at least one card.'),
+        validators.min('Please select at least one card.', 1),
+      ]"
+      :value="selectedCards.length"
+      #="{ props }"
+    >
+      <input-tooltip
+        tooltip="Please select at least one card from the available node cards. Note that if you have a deployment that already uses certain cards, they will not appear in the selection area. You have the option to select one or more cards.."
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/components/select_gpu_node.vue:36]: updated

File: packages/playground/src/components/select_gpu_node.vue [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+watch(selectedNode, async () => {
+  loadingCards.value = true;
+  const projectName = ProjectName.Fullvm.toLowerCase();
+  const grid = await getGrid(profileManager.profile!, projectName);
+  if (grid) {
+    const gpuNodeHelper = new GPUNode(grid);
+    if (selectedNode.value) {
+      gpuNodeHelper.getNodeCards(selectedNode.value).then(res => {
+        nodeCards.value = res.filter(card => card.contract == 0);
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/components/select_gpu_node.vue:121]: Updated

File: packages/playground/src/utils/deploy_vm.ts [hunk 8]
```diff
@@ -42,18 +42,29 @@ async function createMachine(grid: GridClient, machine: Machine, nodePicker: Nod
...
+    //TODO: Remove this condition once this issue is resolved https://github.com/threefoldtech/tfgrid-sdk-ts/issues/703
+    filters.farmId = machine.farmId;
+    filters.farmName = machine.farmName;
+
+    vm.node_id = await nodePicker.pick(await grid.capacity.filterNodes(filters));
+  }
+
   vm.disks = createDisks(machine.disks);
+  vm.gpus = machine.gpus;
+  console.log("vm.node_id: ", vm.node_id);
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/utils/deploy_vm.ts:67]: Removed

File: packages/playground/src/utils/deploy_vm.ts [hunk 9]
```diff
@@ -134,6 +145,10 @@ export interface Machine {
   disks?: Disk[];
   country?: string;
   qsfsDisks?: QsfsDisk[];
+  hasGPU?: boolean;
+  rentedBy?: number;
+  nodeId?: number;
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/utils/deploy_vm.ts:150]: Based on the script in the client https://github.com/threefoldtech/tfgrid-sdk-ts/pull/666/files#diff-75c77c88440bd55706ceb5a41cccb3f087c0b5bb3645fb6cb8435b5576b70b04R24-R26
I have to pass these attrs, and these attrs came from the `full_vm` component, I added them to the machine interface because I need them there.

File: packages/playground/src/utils/filter_node_with_gpu.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        const filters: FilterOptions = {
+          cru: machine.cpu,
+          mru: Math.round(machine.memory / 1024),
+          country: machine.country,
+          hru: machine.qsfsDisks?.reduce((total, disk) => total + disk.cache, 0),
+          sru: machine.disks?.reduce((total, disk) => total + disk.size, machine.rootFilesystemSize || 0),
+          publicIPs: machine.publicIpv4,
+          availableFor: this.grid.twinId,
+          hasGPU: machine.hasGPU,
+          rentedBy: machine.rentedBy,
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/playground/src/utils/filter_node_with_gpu.ts:47]: i passed from the `full_vm` component, = `twin_id`

File: packages/playground/src/weblets/full_vm.vue (hunk 2 above)
Comment (Mahmoud-Emad) [reply packages/playground/src/weblets/full_vm.vue:125]: updated

File: packages/dashboard/src/portal/views/Account.vue [hunk 11]
```diff
@@ -131,9 +131,9 @@ export default class AccountView extends Vue {
           this.$api,
           this.$store.state.credentials.account.address,
         ));
-        const document = await axios.get(this.documentLink);
-        this.documentHash = md5(document.data);
       }
+      const document = await axios.get(this.documentLink);
+      this.documentHash = md5(document.data);
```
[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply packages/dashboard/src/portal/views/Account.vue]: the development branch merged

**Review by AhmedHanafy725 — APPROVED** [review]
